src/realm/object_converter.cpp

```diff
@@ -253,29 +244,335 @@ void InterRealmValueConverter::copy_dictionary(const Obj& src_obj, Obj& dst_obj,
+        SetBasePtr dst = dst_obj.get_setbase_ptr(m_dst_col);
+        copy_set(*src, *dst, update_out);
     }
     else {
         REALM_ASSERT(!m_src_col.is_collection());
-        InterRealmValueConverter::ConversionResult converted_src;
-        if (cmp_src_to_dst(src_obj.get_any(m_src_col), dst_obj.get_any(m_dst_col), &converted_src, update_out)) {
-            if (converted_src.requires_new_embedded_object) {
-                Obj new_embedded = dst_obj.create_and_set_linked_object(m_dst_col);
-                track_new_embedded(converted_src.src_embedded_to_check, new_embedded);
+        // nested collections
+        auto src_mixed = src_obj.get_any(m_src_col);
+        if (src_mixed.is_type(type_List)) {
+            dst_obj.set_collection(m_dst_col, CollectionType::List);
+            Lst<Mixed> src_list{src_obj, m_src_col};
+            Lst<Mixed> dst_list{dst_obj, m_dst_col};
+            handle_list_in_mixed(src_list, dst_list);
+        }
+        else if (src_mixed.is_type(type_Set)) {
+            dst_obj.set_collection(m_dst_col, CollectionType::Set);
+            realm::Set<Mixed> src_set{src_obj, m_src_col};
+            realm::Set<Mixed> dst_set{dst_obj, m_dst_col};
+            // sets cannot be nested, so we just need to copy the values
+            copy_set(src_set, dst_set, nullptr);
+        }
+        else if (src_mixed.is_type(type_Dictionary)) {
+            dst_obj.set_collection(m_dst_col, CollectionType::Dictionary);
+            Dictionary src_dict{src_obj, m_src_col};
+            Dictionary dst_dict{dst_obj, m_dst_col};
+            handle_dictionary_in_mixed(src_dict, dst_dict);
+        }
+        else {
+            InterRealmValueConverter::ConversionResult converted_src;
+            auto dst_mixed = dst_obj.get_any(m_dst_col);
+            if (cmp_src_to_dst(src_mixed, dst_mixed, &converted_src, update_out)) {
+                if (converted_src.requires_new_embedded_object) {
+                    Obj new_embedded = dst_obj.create_and_set_linked_object(m_dst_col);
+                    track_new_embedded(converted_src.src_embedded_to_check, new_embedded);
+                }
+                else {
+                    dst_obj.set_any(m_dst_col, converted_src.converted_value);
+                }
             }
-            else {
-                dst_obj.set_any(m_dst_col, converted_src.converted_value);
+        }
+    }
+}
+
+//
+// Handle collections in mixed. A collection can have N nested levels (expect for Sets). And these levels can be
+// nested in arbitrary way (eg a List within a Dictionary or viceversa). In order to try to merge server changes with
+// client changes, the algorithm needs to go throw each single element in the collection, check its type and perform
+// the most appropriate action in order to miminize the number of notificiations triggered.
+//
+void InterRealmValueConverter::handle_list_in_mixed(const Lst<Mixed>& src_list, Lst<Mixed>& dst_list) const
```

Comment:
It would be super nice to have these functions all handled in the existing algorithm instead of copying the code. I think the main difference is that you chose to handle everything recursively, while the original code adds objects that need to be descended and checked into a list to handle later (embedded objects). You could potentially use the same design here to push all nested collections to a list to check later, that might simplify the code. But this is working as well, so I'll leave it up to you. (not blocking)

Reply:
It is more complicated than that... if I had to merge the 2 logics (embedded links and nested collections), `InterRealmValueConverter::cmp_src_to_dst` would have to change, since we don't have a link type to mark a nested collection, but just the type of the mixed, which could vary again at the next level (whereas for links you are controlling all based on the type). 
Also, in case of mismatches, we should report that immediately in order to avoid going too deep. I have started to design something like this `track_new_embedded` for nested collections, but it is not ready yet.